iree/compiler/Dialect/Stream/Transforms/ScheduleAllocation.cpp

```diff
@@ -0,0 +1,1440 @@
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Analysis/Liveness.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/AsmState.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "iree-stream-schedule-allocation"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Stream {
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Alias analysis
+//===----------------------------------------------------------------------===//
+
+using ValueAliasingMap = llvm::MapVector<Value, SmallPtrSet<Value, 16>>;
+
+// Builds a map of value aliases from aliasee to a set of aliasers.
+// Only values that alias will be present in the map.
+static ValueAliasingMap computeExecutionRegionValueAliases(
+    IREE::Stream::AsyncExecuteOp executeOp) {
+  auto *streamBlock = &executeOp.body().front();
+  ValueAliasingMap valueAliases;
+
+  auto propagateAlias = [&](Value streamValue, Value aliasedValue) {
+    auto &baseSet = valueAliases[streamValue];
+    baseSet.insert(aliasedValue);
+    auto &aliasedSet = valueAliases[aliasedValue];
+    baseSet.insert(aliasedSet.begin(), aliasedSet.end());
+    aliasedSet.insert(streamValue);
+  };
+
+  // Start with outputs so that we handle tied values that may lead all the way
+  // back up the chain to the stream inputs.
+  auto tiedStreamOp =
+      cast<IREE::Util::TiedOpInterface>(executeOp.getOperation());
+  auto yieldOp = cast<IREE::Stream::YieldOp>(streamBlock->getTerminator());
+  for (auto it : llvm::zip(executeOp.results(), yieldOp.operands())) {
+    auto outerResult = std::get<0>(it);
+    auto innerResult = std::get<1>(it);
+    auto tiedOperandIndex =
+        tiedStreamOp.getTiedResultOperandIndex(outerResult.getResultNumber());
+    if (tiedOperandIndex.hasValue()) {
+      auto arg = streamBlock->getArgument(tiedOperandIndex.getValue());
+      propagateAlias(innerResult, arg);
+    }
+  }
+
+  for (auto &op : *streamBlock) {
+    // Tied results reuse their operand buffer.
+    auto tiedOp = dyn_cast<IREE::Util::TiedOpInterface>(op);
+    for (auto result : op.getResults()) {
+      if (!result.getType().isa<IREE::Stream::ResourceType>()) continue;
+      if (tiedOp) {
+        auto tiedOperand = tiedOp.getTiedResultOperand(result);
+        if (tiedOperand) {
+          propagateAlias(result, tiedOperand);
+        }
+      }
+    }
+  }
+
+  // Inverst the value aliaser->aliasee map so that we have for any particular
+  // value the list of all other values that alias it.
```

Comment:
nit: typo

```suggestion
  // Inverse the value aliaser->aliasee map so that we have for any particular
  // value the list of all other values that alias it.
```